File: R/evaluate.R
# =============================================================================
# evaluate.R — Phase 2: Evaluation Pipeline (vitals-based)
# =============================================================================
# Evaluates VLM performance on the Ebbinghaus benchmark using the `vitals`
# framework for structured logging, parallel evaluation, and analysis.
#
# Provides:
#   build_dataset()        — construct a vitals-compatible dataset tibble
#   ebbinghaus_solver()    — custom solver: sends image + text to VLM
#   ebbinghaus_scorer()    — custom scorer: deterministic parse + compare
#   run_evals()            — high-level orchestration across prompts × models
#   fill_prompt()          — resolve prompt template placeholders
#   parse_response()       — extract canonical answer from raw model text
#
# Uses `vitals` for task orchestration and `ellmer` for LLM interaction.
# API keys are read from environment variables by ellmer:
#   ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY
#
# Usage:
#   source("R/evaluate.R")
#
#   trials  <- read.csv("data/trials.csv", stringsAsFactors = FALSE)
#   prompts <- read.csv("data/prompts.csv", stringsAsFactors = FALSE)
#
#   tasks <- run_evals(
#     trials  = trials,
#     prompts = prompts,
#     models  = list(
#       list(provider = "openai",    model = "gpt-5.4-pro"),
#       list(provider = "anthropic", model = "claude-sonnet-4-6")
#     )
#   )
# =============================================================================

library(vitals)
library(ellmer)
library(tibble)
library(purrr)

source("R/strip_answer.R")

# =============================================================================
# Direction helpers
# =============================================================================

DIRECTION_WORDS <- list(
  horizontal = list(a = "left",       b = "right"),
  vertical   = list(a = "top",        b = "bottom"),
  diagonal   = list(a = "upper-left", b = "lower-right")
)

#' Fill prompt template placeholders for a specific trial row.
#'
#' Replaces `{direction_a}`, `{direction_b}`, `{test_a_shape}`, etc. with
#' values from the trial. Unknown placeholders are left as-is.
#'
#' @param template Character string. The `user_prompt_template` from prompts.csv.
#' @param trial    One-row data frame or named list from trials.csv.
#' @return Filled prompt string.
fill_prompt <- function(template, trial) {
  orientation <- trial[["orientation"]]
  dirs <- DIRECTION_WORDS[[orientation]]
  if (is.null(dirs)) stop("Unknown orientation: ", orientation)

  out <- template
  out <- gsub("{direction_a}", dirs$a, out, fixed = TRUE)
  out <- gsub("{direction_b}", dirs$b, out, fixed = TRUE)
  out <- gsub("{test_a_shape}", trial[["test_a_shape"]], out, fixed = TRUE)
  out <- gsub("{test_b_shape}", trial[["test_b_shape"]], out, fixed = TRUE)
  out
}

# =============================================================================
# Response parser
# =============================================================================

#' Extract a structured response_larger value from raw model text.
#'
#' Maps directional words (left/right/top/bottom/upper-left/lower-right) to
#' "a" or "b" based on the trial's orientation, then normalises to the
#' canonical set: "a", "b", "equal", "unknown", "parse_error".
#'
#' For chain-of-thought prompts the function first looks for an explicit
#' "ANSWER: <word>" tag; if absent it falls back to scanning the full text.
#'
#' @param raw_response Character. Full verbatim model response.
#' @param orientation  Character. Trial orientation ("horizontal", "vertical",
#'   "diagonal").
#' @param response_format Character. Prompt's response_format field
#'   ("forced_choice" or "free_text").
#' @return One of: "a", "b", "equal", "unknown", "parse_error".
parse_response <- function(raw_response, orientation, response_format = "forced_choice") {
  if (is.null(raw_response) || is.na(raw_response) || nchar(trimws(raw_response)) == 0) {
    return("parse_error")
  }

  dirs <- DIRECTION_WORDS[[orientation]]
  if (is.null(dirs)) stop("Unknown orientation: ", orientation)

  # For chain-of-thought, prefer the explicit ANSWER tag
  text <- raw_response
  if (response_format == "free_text") {
    answer_match <- regmatches(text, regexpr("(?i)ANSWER\\s*:\\s*(\\S+)", text, perl = TRUE))
    if (length(answer_match) > 0 && nchar(answer_match) > 0) {
      text <- sub("(?i).*ANSWER\\s*:\\s*(\\S+).*", "\\1", answer_match, perl = TRUE)
    }
  }

  # Normalise: lowercase, strip punctuation (keep hyphens and spaces), split words
  text_clean <- tolower(trimws(text))
  text_clean <- gsub("[^a-z -]", " ", text_clean)
  text_clean <- trimws(text_clean)
  words <- strsplit(text_clean, "\\s+")[[1]]
  first_word <- words[1]
  # For compound directions like "upper left" -> "upper-left"
  first_two <- if (length(words) >= 2) paste(words[1:2], collapse = "-") else ""

  # Map to canonical answer
  if (is.na(first_word) || first_word == "") return("parse_error")

  if (first_word == dirs$a || first_two == dirs$a || first_word == "a") return("a")
  if (first_word == dirs$b || first_two == dirs$b || first_word == "b") return("b")
  if (first_word %in% c("equal", "same", "neither")) return("equal")
  if (first_word %in% c("unknown", "unsure", "unclear", "cannot", "can't", "not")) return("unknown")

  # Fallback: scan full cleaned text for any direction word (try both hyphenated and spaced)
  full_clean <- tolower(gsub("[^a-z -]", " ", raw_response))
  dir_a_pat <- paste0(gsub("-", "[ -]", dirs$a, fixed = TRUE))
  dir_b_pat <- paste0(gsub("-", "[ -]", dirs$b, fixed = TRUE))
  if (grepl(dir_a_pat, full_clean)) return("a")
  if (grepl(dir_b_pat, full_clean)) return("b")
  if (grepl("equal|same|neither", full_clean))  return("equal")
  if (grepl("unknown|unsure|unclear",  full_clean)) return("unknown")

  "parse_error"
}

# =============================================================================
# Dataset builder
# =============================================================================

#' Build a vitals-compatible dataset from trials and a single prompt variant.
#'
#' Each row represents one (trial, prompt) pair. The `input` column contains
#' the filled user prompt, `target` contains the ground truth answer, and
#' metadata columns carry everything needed by the solver and scorer.
#'
#' @param trials    Data frame from trials.csv.
#' @param prompt    One-row data frame from prompts.csv.
#' @param image_dir Directory with answer-stripped images.
#' @return A tibble suitable for `Task$new(dataset = ...)`.
build_dataset <- function(trials, prompt, image_dir = "images_eval") {
  tibble(
    id              = as.character(trials$trial_id),
    input           = vapply(
      seq_len(nrow(trials)),
      function(i) fill_prompt(prompt$user_prompt_template, trials[i, ]),
      character(1)
    ),
    target          = trials$true_larger,
    # Metadata for solver and scorer
    trial_id        = trials$trial_id,
    orientation     = trials$orientation,
    tier            = trials$tier,
    true_diff_ratio = trials$true_diff_ratio,
    prompt_id       = prompt$prompt_id,
    response_format = prompt$response_format,
    image_path      = file.path(
      image_dir,
      paste0(trials$trial_id, ".", trials$file_format)
    )
  )
}

# =============================================================================
# Custom solver
# =============================================================================

#' Create an Ebbinghaus solver that sends image + text to a VLM.
#'
#' Returns a solver function compatible with vitals `Task$new(solver = ...)`.
#' The solver reads image paths from the dataset (passed via `image_paths`
#' argument) and sends each image alongside the text prompt to the model.
#'
#' @param solver_chat An ellmer Chat object (will be cloned per sample).
#' @return A solver function.
ebbinghaus_solver <- function(inputs, ..., solver_chat, image_paths) {

  # Build one chat clone per sample and send image + text
  chats <- map2(inputs, image_paths, function(prompt_text, img_path) {
    ch <- solver_chat$clone()
    ch$chat(prompt_text, content_image_file(img_path, resize = "none"))
    ch
  })

  list(
    result      = map_chr(chats, function(ch) ch$last_turn()@text),
    solver_chat = chats
  )
}

# =============================================================================
# Custom scorer
# =============================================================================

#' Score Ebbinghaus evaluation results using deterministic parsing.
#'
#' Parses each raw model response using `parse_response()`, compares to
#' `target`, and returns an ordered factor (I/C). Stores parsed responses
#' in `scorer_metadata`.
#'
#' @param samples Tibble from `task$get_samples()` — includes `result`,
#'   `target`, `orientation`, `response_format`.
#' @return A list with `score` (ordered factor) and `scorer_metadata`.
ebbinghaus_scorer <- function(samples) {

  parsed <- pmap_chr(
    list(samples$result, samples$orientation, samples$response_format),
    function(result, orientation, response_format) {
      parse_response(result, orientation, response_format)
    }
  )

  correct <- (parsed == samples$target)

  scores <- factor(
    ifelse(correct, "C", "I"),
    levels  = c("I", "C"),
    ordered = TRUE
  )

  list(
    score           = scores,
    scorer_metadata = tibble(
      parsed_response = parsed,
      correct         = correct
    )
  )
}

# =============================================================================
# Chat constructor helper
# =============================================================================

#' Create an ellmer Chat for a model config + system prompt.
#'
#' @param model_cfg Named list with `provider`, `model`, and optionally
#'   `temperature`, `max_tokens`.
#' @param system_prompt Character: system prompt string.
#' @return An ellmer Chat object.
make_chat <- function(model_cfg, system_prompt) {
  p <- params(
    temperature = model_cfg$temperature %||% 0,
    max_tokens  = model_cfg$max_tokens  %||% 512L
  )

  switch(model_cfg$provider,
    anthropic = chat_anthropic(
      system_prompt = system_prompt, model = model_cfg$model, params = p
    ),
    openai = chat_openai(
      system_prompt = system_prompt, model = model_cfg$model, params = p
    ),
    google = chat_google_gemini(
      system_prompt = system_prompt, model = model_cfg$model, params = p
    ),
    github = chat_github(
      system_prompt = system_prompt, model = model_cfg$model, params = p
    ),
    stop("Unknown provider: ", model_cfg$provider,
         ". Use 'anthropic', 'openai', 'google', or 'github'.")
  )
}

# =============================================================================
# High-level orchestration
# =============================================================================

#' Run the full evaluation across prompts × models.
#'
#' Creates one vitals Task per (prompt, model) combination, evaluates it,
#' and returns a named list of Task objects. Use `vitals_bind()` on the
#' result to get a combined data frame for analysis.
#'
#' @param trials    Data frame from trials.csv.
#' @param prompts   Data frame from prompts.csv.
#' @param models    List of named lists. Each must have `provider` and `model`;
#'   optionally `temperature`, `max_tokens`. Example:
#'   `list(provider = "openai", model = "gpt-5.4-pro")`.
#' @param image_dir Directory containing answer-stripped images. If the
#'   directory doesn't exist, images will be stripped from `images/`.
#' @param epochs    Number of times to repeat each sample (default 1).
#' @param view      Open the Inspect log viewer after each eval (default:
#'   only at the end, not after each task).
#'
#' @return A named list of Task objects. Names follow the pattern
#'   `"<provider>/<model>__<prompt_description>"`.
run_evals <- function(trials,
                      prompts,
                      models,
                      image_dir = "images_eval",
                      epochs    = 1L,
                      view      = FALSE) {

  # Strip answers from images if needed
  if (!dir.exists(image_dir) || length(list.files(image_dir)) == 0) {
    message("Stripping answers from images...")
    strip_answer_from_images(trials, image_dir)
  }

  tasks <- list()

  for (mi in seq_along(models)) {
    mcfg <- models[[mi]]
    model_label <- paste0(mcfg$provider, "/", mcfg$model)

    for (p_idx in seq_len(nrow(prompts))) {
      prompt <- prompts[p_idx, ]
      task_name <- paste0(model_label, "__", prompt$description)

      message(sprintf("Creating task: %s", task_name))

      dataset <- build_dataset(trials, prompt, image_dir)
      chat    <- make_chat(mcfg, prompt$system_prompt)

      task <- Task$new(
        dataset = dataset,
        solver  = ebbinghaus_solver,
        scorer  = ebbinghaus_scorer,
        epochs  = epochs,
        name    = task_name
      )

      message(sprintf("Evaluating: %s", task_name))

      task$eval(
        solver_chat = chat,
        image_paths = dataset$image_path,
        view        = view
      )

      tasks[[task_name]] <- task
    }
  }

  message("All evaluations complete. ", length(tasks), " tasks evaluated.")
  message("Use vitals_view() to open the Inspect log viewer.")

  invisible(tasks)
}

# =============================================================================
# Results extraction
# =============================================================================

#' Extract evaluation results from tasks into a flat data frame.
#'
#' Combines all Task objects via `vitals_bind()` and unnests metadata
#' columns needed for downstream analysis. This is the primary bridge
#' between vitals output and `analyze_results()`.
#'
#' @param tasks Named list of Task objects (as returned by `run_evals()`).
#' @return A tibble with columns: task, id, score, trial_id, orientation,
#'   tier, true_diff_ratio, prompt_id, parsed_response, correct,
#'   model_label, prompt_description.
extract_results <- function(tasks) {
  bound <- do.call(vitals_bind, tasks)

  # Parse task name to extract model and prompt info
  bound$model_label <- sub("__.*$", "", bound$task)
  bound$prompt_description <- sub("^.*__", "", bound$task)

  bound
}

File: py/src/evaluate.py
"""
evaluate.py — Phase 2: Evaluation Pipeline (Inspect AI-based)

Evaluates VLM performance on the Ebbinghaus benchmark using Inspect AI
for structured logging, evaluation, and analysis.

Provides:
  build_dataset()         — construct Inspect AI samples from trials
  ebbinghaus_solver()     — custom solver: sends image + text to VLM
  ebbinghaus_scorer()     — custom scorer: deterministic parse + compare
  run_evals()             — high-level orchestration across prompts × models
  fill_prompt()           — resolve prompt template placeholders
  parse_response()        — extract canonical answer from raw model text

Uses Inspect AI for task orchestration. API keys are read from env vars:
  ANTHROPIC_API_KEY, OPENAI_API_KEY, GOOGLE_API_KEY

Usage:
  from src.evaluate import run_evals
  import pandas as pd

  trials  = pd.read_csv("data/trials.csv")
  prompts = pd.read_csv("data/prompts.csv")

  logs = run_evals(
      trials  = trials,
      prompts = prompts,
      models  = [
          {"provider": "openai",    "model": "gpt-5.4-pro"},
          {"provider": "anthropic", "model": "claude-sonnet-4-6"},
      ],
  )
"""

import os
import re
import pandas as pd
from inspect_ai import Task, eval as inspect_eval
from inspect_ai.dataset import Sample, MemoryDataset
from inspect_ai.model import ChatMessageUser, ContentImage, ContentText, get_model
from inspect_ai.scorer import Score, scorer, Target, CORRECT, INCORRECT
from inspect_ai.solver import Generate, TaskState, solver

from src.strip_answer import strip_answer_from_images


# =============================================================================
# Direction helpers
# =============================================================================

DIRECTION_WORDS = {
    "horizontal": {"a": "left",       "b": "right"},
    "vertical":   {"a": "top",        "b": "bottom"},
    "diagonal":   {"a": "upper-left", "b": "lower-right"},
}


def fill_prompt(template: str, trial: dict) -> str:
    """Fill prompt template placeholders for a specific trial.

    Replaces {direction_a}, {direction_b}, {test_a_shape}, {test_b_shape}
    with values from the trial.
    """
    orientation = trial["orientation"]
    dirs = DIRECTION_WORDS.get(orientation)
    if dirs is None:
        raise ValueError(f"Unknown orientation: {orientation}")

    out = template
    out = out.replace("{direction_a}", dirs["a"])
    out = out.replace("{direction_b}", dirs["b"])
    out = out.replace("{test_a_shape}", str(trial["test_a_shape"]))
    out = out.replace("{test_b_shape}", str(trial["test_b_shape"]))
    return out


# =============================================================================
# Response parser
# =============================================================================

def parse_response(
    raw_response: str,
    orientation: str,
    response_format: str = "forced_choice",
) -> str:
    """Extract a structured response_larger value from raw model text.

    Maps directional words to "a" or "b" based on the trial's orientation,
    then normalises to: "a", "b", "equal", "unknown", "parse_error".
    """
    if not raw_response or not raw_response.strip():
        return "parse_error"

    dirs = DIRECTION_WORDS.get(orientation)
    if dirs is None:
        raise ValueError(f"Unknown orientation: {orientation}")

    # For chain-of-thought, prefer the explicit ANSWER tag
    text = raw_response
    if response_format == "free_text":
        match = re.search(r"(?i)ANSWER\s*:\s*(\S+)", text)
        if match:
            text = match.group(1)

    # Normalise: lowercase, strip punctuation (keep hyphens and spaces)
    text_clean = re.sub(r"[^a-z \-]", " ", text.lower().strip()).strip()
    words = text_clean.split()
    first_word = words[0] if words else ""
    first_two = "-".join(words[:2]) if len(words) >= 2 else ""

    if not first_word:
        return "parse_error"

    if first_word == dirs["a"] or first_two == dirs["a"] or first_word == "a":
        return "a"
    if first_word == dirs["b"] or first_two == dirs["b"] or first_word == "b":
        return "b"
    if first_word in ("equal", "same", "neither"):
        return "equal"
    if first_word in ("unknown", "unsure", "unclear", "cannot", "can't", "not"):
        return "unknown"

    # Fallback: scan full text
    full_clean = re.sub(r"[^a-z \-]", " ", raw_response.lower())
    dir_a_pat = dirs["a"].replace("-", "[ -]")
    dir_b_pat = dirs["b"].replace("-", "[ -]")
    if re.search(dir_a_pat, full_clean):
        return "a"
    if re.search(dir_b_pat, full_clean):
        return "b"
    if re.search(r"equal|same|neither", full_clean):
        return "equal"
    if re.search(r"unknown|unsure|unclear", full_clean):
        return "unknown"

    return "parse_error"


# =============================================================================
# Dataset builder
# =============================================================================

def build_dataset(
    trials: pd.DataFrame, prompt: pd.Series, image_dir: str = "images_eval"
) -> list[Sample]:
    """Build Inspect AI samples from trials and a single prompt variant.

    Each Sample represents one (trial, prompt) pair.
    """
    samples = []
    for i in range(len(trials)):
        row = trials.iloc[i]
        filled = fill_prompt(prompt["user_prompt_template"], row.to_dict())
        img_path = os.path.join(
            image_dir, f"{int(row['trial_id'])}.{row['file_format']}"
        )

        samples.append(Sample(
            input=[ChatMessageUser(content=[
                ContentImage(image=img_path),
                ContentText(text=filled),
            ])],
            target=row["true_larger"],
            id=str(int(row["trial_id"])),
            metadata={
                "trial_id": int(row["trial_id"]),
                "orientation": row["orientation"],
                "tier": int(row["tier"]) if pd.notna(row["tier"]) else None,
                "true_diff_ratio": row["true_diff_ratio"],
                "prompt_id": prompt["prompt_id"],
                "response_format": prompt["response_format"],
            },
        ))

    return samples


# =============================================================================
# Custom solver
# =============================================================================

@solver
def ebbinghaus_solver():
    """Solver that passes through the image + text to the model."""
    async def solve(state: TaskState, generate: Generate) -> TaskState:
        return await generate(state)
    return solve


# =============================================================================
# Custom scorer
# =============================================================================

@scorer(metrics=["accuracy"])
def ebbinghaus_scorer():
    """Score Ebbinghaus evaluation results using deterministic parsing."""
    async def score(state: TaskState, target: Target) -> Score:
        raw = state.output.completion
        orientation = state.metadata.get("orientation", "horizontal")
        response_format = state.metadata.get("response_format", "forced_choice")

        parsed = parse_response(raw, orientation, response_format)
        correct = parsed == target.text

        return Score(
            value=CORRECT if correct else INCORRECT,
            answer=parsed,
            metadata={
                "parsed_response": parsed,
                "correct": correct,
            },
        )
    return score


# =============================================================================
# High-level orchestration
# =============================================================================

def run_evals(
    trials: pd.DataFrame,
    prompts: pd.DataFrame,
    models: list[dict],
    image_dir: str = "images_eval",
    epochs: int = 1,
) -> list:
    """Run the full evaluation across prompts × models.

    Creates one Inspect AI Task per (prompt, model) combination, evaluates it,
    and returns a list of eval logs.

    Args:
        trials: DataFrame from trials.csv.
        prompts: DataFrame from prompts.csv.
        models: List of dicts with 'provider', 'model', optionally
            'temperature', 'max_tokens'.
        image_dir: Directory containing answer-stripped images.
        epochs: Number of times to repeat each sample.

    Returns:
        A list of Inspect AI eval log objects.
    """
    # Strip answers from images if needed
    if not os.path.isdir(image_dir) or not os.listdir(image_dir):
        print("Stripping answers from images...")
        strip_answer_from_images(trials, image_dir)

    all_logs = []

    for mcfg in models:
        model_label = f"{mcfg['provider']}/{mcfg['model']}"
        model_str = f"{mcfg['provider']}/{mcfg['model']}"

        for p_idx in range(len(prompts)):
            prompt = prompts.iloc[p_idx]
            task_name = f"{model_label}__{prompt['description']}"
            print(f"Creating task: {task_name}")

            samples = build_dataset(trials, prompt, image_dir)
            dataset = MemoryDataset(samples=samples, name=task_name)

            task = Task(
                dataset=dataset,
                solver=ebbinghaus_solver(),
                scorer=ebbinghaus_scorer(),
                epochs=epochs,
            )

            print(f"Evaluating: {task_name}")

            model_kwargs = {}
            if "temperature" in mcfg:
                model_kwargs["temperature"] = mcfg["temperature"]
            if "max_tokens" in mcfg:
                model_kwargs["max_tokens"] = mcfg["max_tokens"]

            logs = inspect_eval(
                task,
                model=get_model(model_str, **model_kwargs),
            )
            all_logs.extend(logs)

    print(f"All evaluations complete. {len(all_logs)} logs generated.")
    return all_logs
